homeassistant/components/soma/cover.py

```diff
@@ -0,0 +1,81 @@
+"""
+Support for Soma Covers.
+
+For more details about this platform, please refer to the documentation at
+"""
+
+import logging
+
+from homeassistant.components.cover import CoverDevice, ATTR_POSITION
+from homeassistant.components.soma import DOMAIN, SomaEntity, DEVICES, API
+from homeassistant.exceptions import PlatformNotReady
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Soma cover platform."""
+    def get_covers():
+        """Retrieve covers."""
+        devices = hass.data[DOMAIN][DEVICES]
+
+        return [SomaCover(cover, hass.data[DOMAIN][API]) for cover in
+                devices]
+
+    async_add_entities(await hass.async_add_executor_job(get_covers), True)
```

Comment:
Is something in `get_covers` doing I/O?

Reply:
No, I changed this to add_entities and it seems to work. Is this what you had in mind?

Reply:
If it's just in memory operations we can create the entities and call `async_add_entities` directly in `async_setup_entry`. We don't need the `get_covers` function.

Reply:
Fixed